### PR TITLE
many: when building initramfs, pull files from the run system

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Previously we were building a rootfs from files at debian package build time, and the ubuntu-core-initramfs was generating the initramfs from this file. This PR changes this script so it pulls files from other packages when invoked, instead. This removes most of the build dependencies and makes maintenance of the package easier, as it does not need to be updated when systemd or other run-time dependencies are updated anymore.

### CODE REVIEW

**Review by xnox — CHANGES_REQUESTED** [review]
Did you create initrd before and after the change and do they look the same?

or would you like me to build multiple kernel snaps side by side, for us to compare?

File: bin/ubuntu-core-initramfs [hunk 1]
```diff
@@ -323,15 +488,10 @@ def create_initrd(parser, args):
...
-                [
-                    "/usr/lib/dracut/dracut-install",
-                    "-D", main,
-                    "--ldd", "--all",
+            install_files([
                     "/usr/bin/kcapi-hasher",
                     "/usr/bin/.kcapi-hasher.hmac",
-                ] + glob.glob("/usr/lib/*/.libkcapi.so.*.hmac")
-            )
+            ] + glob.glob("/usr/lib/*/.libkcapi.so.*.hmac"), main)
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply bin/ubuntu-core-initramfs:494]: I've added a TODO about this

**Review by alfonsosanchezbeato — COMMENTED** [review]
@xnox thanks for the quick review, I've addressed your comments now.

FTR, yes, to check correctness I have compared the final list of files with what the old approach produced. The are just some small differences that should be fine:
1. Some links/binaries that used to be in `/usr/bin` are now in `/usr/sbin`: reboot, rmmod, e2fsck and others, which seems more correct to me as that is their path in the rootfs. Not sure how they were originally placed in `/usr/bin`.
2. Due to an error in an invocation of dracut-install, we were pulling more libraries than needed in the initramfs (concretely, when calling it on `/usr/lib/$(DEB_HOST_MULTIARCH)/plymouth/*.so`, where we should have run it only for the plugins we are pulling - two-step, script, label-ft). I've fixed this so these unneeded libraries are not included anymore.

Fwiw, I've tested this locally by replacing initramfs in a 24 snap kernel and things were good, also CI seems to be happy.

**Comment by alfonsosanchezbeato** [comment]
Hm, ubuntu-image in edge is failing now...

**Comment by alfonsosanchezbeato** [comment]
There is an issue with the just merged assertions, see https://github.com/snapcore/models/pull/17#issuecomment-1836612194

File: debian/rules [hunk 2]
```diff
@@ -66,61 +21,9 @@ override_dh_auto_install:
...
-		ln -v -s busybox debian/tmp/usr/bin/$$alias; \
-	done
-
-	LD_PRELOAD= /usr/lib/dracut/dracut-install -D $(CURDIR)/debian/tmp --ldd $(INSTALL_FILES_FROM_HOST)
-	dpkg -L dmsetup | grep rules.d | xargs -L1 /usr/lib/dracut/dracut-install -D $(CURDIR)/debian/tmp --ldd
-	ln -v -s e2fsck debian/tmp/usr/sbin/fsck.ext4
-
-	set -e; \
-	for e in $$(find debian/tmp -type f -executable) \
-		/usr/lib/$(DEB_HOST_MULTIARCH)/plymouth/*.so \
```
Comment (valentindavid) [inline debian/rules:111]: It seems we had a bug here where would install all dependencies of `label.so`. Now we do not. That is good. :+1:

**Review by valentindavid — COMMENTED** [review]
Looks good. Some minor comments. The `$$` is the only one that requires a fix before merge.

I would have kept `/usr/sbin` as symlink change it in a follow up PR so we do not have to revert the whole thing if we find an issue. But maybe it is safe enough.

File: bin/ubuntu-core-initramfs [hunk 3]
```diff
@@ -272,7 +272,192 @@ def add_modules_from_file(dest_d, kernel_root, modules_d, fw_d, conf_file, db,
...
+                           ".*systemd-pcrphase.*")
+    filtered = [i for i in files if not to_remove.match(i)]
+    # Install
+    install_files(filtered, dest_dir)
+
+    # Local modifications
+    # This hack should be removed with PR#113
+    check_call([r"sed -i '/^After=/"
+                r"{;s, *plymouth-start[.]service *, ,;/"
+                r"^After= *$$/d;}' "
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply bin/ubuntu-core-initramfs]: Good catch, extra `$` removed now

File: bin/ubuntu-core-initramfs [hunk 4]
```diff
@@ -272,7 +272,192 @@ def add_modules_from_file(dest_d, kernel_root, modules_d, fw_d, conf_file, db,
...
+                + os.path.join(dest_dir,
+                               "usr/lib/systemd/system/systemd-ask-password-*")],
+               shell=True)
+    # Generate hw database (/usr/lib/udev/hwdb.bin) for udev and
+    # remove redundant definitions after that.
+    check_call(["systemd-hwdb", "--root", dest_dir,
+                "update", "--usr", "--strict"])
+    check_call(["rm -rf " + os.path.join(dest_dir,
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply bin/ubuntu-core-initramfs]: Changed

File: bin/ubuntu-core-initramfs [hunk 5]
```diff
@@ -272,7 +272,192 @@ def add_modules_from_file(dest_d, kernel_root, modules_d, fw_d, conf_file, db,
...
+
+def install_busybox(dest_dir):
+    install_file_to_path("/usr/lib/initramfs-tools/bin/busybox", dest_dir,
+                         "usr/bin/busybox")
+    # Create links to commands
+    bb_cmd = os.path.join(dest_dir, "usr/bin/busybox")
+    out = check_output([bb_cmd, "--list-long"])
+    cmds = out.decode("utf-8").splitlines()
+    # Remove command we do not want
+    to_remove = ["busybox", "reboot", "mount", "umount"]
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply bin/ubuntu-core-initramfs]: Excellent point, I've removed it explicitly (it was still in usr/bin, although the one in usr/sbin already pointed to kmod)

File: debian/rules (hunk 2 above)
Comment (alfonsosanchezbeato) [reply debian/rules:111]: Correct, see also my comment to Dimitri.

**Review by alfonsosanchezbeato — COMMENTED** [review]
@valentindavid thanks for the review, comments addressed now

**Comment by alfonsosanchezbeato** [comment]
@valentindavid @xnox this is ready for another review round.

**Review by valentindavid — APPROVED** [review]
Looks good. Thank you!

**Review by xnox — APPROVED** [review]
Here be dragons!

**Pull request merged**